treq/testing.py

```diff
@@ -205,3 +207,212 @@ def __init__(self, resource):
+    """
+    def __repr__(self):
+        return "ANYTHING"
+
+    def __eq__(self, _):
+        return True
+
+    def __ne__(self, _):
+        return False
+
+ANY = _any()
+
+
+class SequenceStringStubs(object):
+    """
+    Takes a sequence of::
+
+        [((method, url, params, headers, data), (code, headers, body)),
+         ...]
+
+    Expects the requests to arrive in sequence order.  If there are no more
+    responses, or the request's paramters do not match the next item's expected
+    request paramters, raises :obj:`AssertionError`.
+
+    :ivar list sequence: The sequence of expected request arguments mapped to
+        stubbed responses
+    :ivar testcase: A :obj:`TestCase` that can be used to report failures in
+        matching requests against expected requests.  Raising exceptions does
+        not work, because :obj:`Resource.render` will just convert that into
+        a 500 response.
+    """
+    def __init__(self, sequence, testcase):
+        self._sequence = sequence
+        self._testcase = testcase
+
+    def consumed(self):
+        """
+        :return: `bool` representing whether the entire sequence has been
+            consumed.  This is useful in tests to assert that the expected
+            requests have all been made.
+        """
+        return len(self._sequence) == 0
+
+    @contextmanager
+    def consume(self):
+        """
+        Usage::
+
+            stub_treq = StubTreq(StringStubbingResource(
+                sequence_stubs.get_response_for))
+            with sequence_stubs.consume():
+                stub_treq.get('http://fakeurl.com')
+                stub_treq.get('http://another-fake-url.com')
+
+        If there are still remaining expected requests to be made in the
+        sequence, fails the provided test case.
+
+        :return: a context manager that can be used to ensure all expected
+            requests have been made.
+        """
+        yield
+        if not self.consumed():
+            self._testcase.fail("\n".join(
+                ["Not all expected requests were made.  Still expecting:"] +
+                ["- {0}(url={1}, params={2}, headers={3}, data={4})".format(
+                    *expected) for expected, _ in self._sequence]))
+
+    def get_response_for(self, method, url, params, headers, data):
+        """
+        :return: the next response in the sequence, provided that the
+            parameters match the next in the sequence.
+        """
+        if len(self._sequence) == 0:
+            self._testcase.addCleanup(
+                self._testcase.fail,
+                "No more requests expected, but request {0!r} made.".format(
+                    (method, url, params, headers, data)))
+            return (500, {}, "StubbingError")
+
+        expected, response = self._sequence[0]
+        e_method, e_url, e_params, e_headers, e_data = expected
+
+        checks = [
+            (e_method is ANY or e_method.lower() == method.lower(), "method"),
```

Comment:
> Just to clarify, you mean just for the method right? The comment above also mentioned about the headers, and while I agree with you that the ANY checks everywhere is weird, I'd also like it to be consistent.

Yes, the comment you replied to was just about the `e_method` check.

Okay, on to your other points...

First, I'm going to assume when you say "parameters" in this comment that you're talking about the items given in the tuples passed to `SequenceStringStubs`, not the parameters to `get_response_for`. 

> It seems necessary for params, because there needs to be some kind of way to distinguish between no params and any params

I read "It" as "explicitly checking `ANY`" here, but now I'm confused because we're _not_ explicitly checking `ANY` against `e_params`. If you meant headers instead of params (just a shot in the dark), then I still don't understand because the current code already doesn't give you a way to check "no headers" vs "any headers" -- passing `{}` or `ANY` both mean the same thing, which is "any headers", since we hardcode the use of `HasHeaders`.

> And its' easier to document that ANY can be used for all these values, instead of "the first parameter is required, the second and third parameters can be ANY, but for the fourth parameter {} represents ANY, and the fifth parameter can be ANY.

To pick on the point about parameter #4: it's not that `{}` "represents" `ANY` -- `ANY` is a just an unnecessary concept here since we've already got a perfectly good value that will always be a subset of the actual headers, and that is `{}`.

If we use `ANY` as the de facto sentinel, then okay, it makes sense to allow it for any element. But I'm trying to figure out a way to eradicate the explicit checks against `ANY` entirely.

>  (speaking of which do you think I should also accept None?)

I think this would be good, but I'd like to see if we can have `None` replace `ANY` entirely to mean "don't perform this check". That seems like it would be more idiomatic/unsurprising than giving `ANY` some special meaning. And I don't _think_ there's any overloading problems there, assuming that all the things we're checking have a reasonable "empty" value, which I'm pretty sure is true: params is `{}`, headers is `{}`, data is `""`. Or is there a meaningful distinction between a `data` of `""` and `data` of `None`?


## The important part of this comment:

(I considered just deleting everything I'd written before this part because I think this part pretty succinctly covers my position, but I'll leave that stuff there for posterity, and in case my proposal doesn't work out ;-)

I think it would be even nicer if we could avoid having a sentinel at all and just use ANY "naturally" for equality checks. I'm defining "naturally" as "only using it for its equality semantics". Lemme think about each field:

1. method - no problem, `ANY` can be used naturally
2. URL - URLPath has no equality method. wtf. and even if it did it may be annoying to require users to construct URLPath objects instead of just passing strings. Oh! But wait! Why can't we just make the check `e_url == str(URLPath.fromString(url))`? That would allow users to use `ANY` naturally. Why were we round-tripping the expected URL through URLPath?
3. params - no problem, users can use ANY naturally.
4. headers - since we're not using equality here, ANY is inappropriate, but `{}` has the desired meaning so users can just pass that.
5. data - no problem, `ANY` can be used naturally.


Okay, so unless there's something wrong with my proposal to not round-trip through URLPath, I think we can get rid of *any* use of an explicit "don't check this field" sentinel. And I don't think there's a documentation problem here -- we must just document that each field must be given. A friendly hint that `ANY` is a good way to not care about particular fields at the end of the docstring should be good enough.

But I'm sure I'm missing something :)